app/models/external_activity.rb

```diff
@@ -207,6 +207,80 @@ def duplicateable?(user)
     false
   end
 
+  # Duplicates external activity both locally and on the external authoring system (e.g. LARA).
+  # New activity will have a new owner and publication status set to private.
+  # Returns a new activity or nil in case of error.
+  def duplicate(new_owner, root_url)
+    clone = ExternalActivity.new(attributes.except('id', 'uuid', 'created_at', 'updated_at'))
+    clone.name = "Copy of #{name}"
+    clone.user = new_owner
+    clone.publication_status = 'private'
+    clone.save
+
+    # This section triggers remote duplication if necessary.
+    # Do it as soon as possible, as it includes communication with external system. In case of failure,
+    # we can cancel this action earlier, before we create bunch of intermediate objects that would need to be
+    # cleaned up too.
+    if source_type === 'LARA'
+      unless clone.duplicate_on_lara(root_url)
+        # If duplication on LARA fails, destroy clone and return nil.
+        clone.destroy
+        return nil
+      end
+    end
+
+    # Copy rest of the activity properties.
+    clone.material_property_list = material_property_list
+    clone.grade_level_list = grade_level_list
+    clone.subject_area_list = subject_area_list
+    clone.sensor_list = sensor_list
+    # Copy projects. Limit projects to ones that can be assigned by the new author.
+    allowed_projects = Admin::Project.all.select{ |p| Admin::ProjectPolicy.new(new_owner, p).assign_to_material? }.map(&:id)
+    clone.project_ids = project_ids & allowed_projects
```

Comment:
How about replacing `Admin::Project.all` with `Admin::Project.where(id: project_ids)`?

Then you can drop line 239 and we don't need to load in every project in the database.

Reply:
Thanks, good point, done.